{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samples</th>\n",
       "      <th>type</th>\n",
       "      <th>1007_s_at</th>\n",
       "      <th>1053_at</th>\n",
       "      <th>117_at</th>\n",
       "      <th>121_at</th>\n",
       "      <th>1255_g_at</th>\n",
       "      <th>1294_at</th>\n",
       "      <th>1316_at</th>\n",
       "      <th>1320_at</th>\n",
       "      <th>...</th>\n",
       "      <th>AFFX-r2-Ec-bioD-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioD-5_at</th>\n",
       "      <th>AFFX-r2-P1-cre-3_at</th>\n",
       "      <th>AFFX-r2-P1-cre-5_at</th>\n",
       "      <th>AFFX-ThrX-3_at</th>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>136</td>\n",
       "      <td>tumoral</td>\n",
       "      <td>8.767514</td>\n",
       "      <td>4.092134</td>\n",
       "      <td>6.006339</td>\n",
       "      <td>6.091951</td>\n",
       "      <td>2.738641</td>\n",
       "      <td>7.707960</td>\n",
       "      <td>6.702969</td>\n",
       "      <td>5.203349</td>\n",
       "      <td>...</td>\n",
       "      <td>13.852908</td>\n",
       "      <td>13.854510</td>\n",
       "      <td>14.953006</td>\n",
       "      <td>14.824910</td>\n",
       "      <td>3.710845</td>\n",
       "      <td>2.493435</td>\n",
       "      <td>2.913242</td>\n",
       "      <td>2.360155</td>\n",
       "      <td>2.758725</td>\n",
       "      <td>2.779035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>137</td>\n",
       "      <td>tumoral</td>\n",
       "      <td>9.177852</td>\n",
       "      <td>5.688045</td>\n",
       "      <td>5.599253</td>\n",
       "      <td>5.943833</td>\n",
       "      <td>2.667376</td>\n",
       "      <td>6.737484</td>\n",
       "      <td>7.336429</td>\n",
       "      <td>6.246692</td>\n",
       "      <td>...</td>\n",
       "      <td>13.784316</td>\n",
       "      <td>13.776588</td>\n",
       "      <td>14.944678</td>\n",
       "      <td>14.798245</td>\n",
       "      <td>3.840439</td>\n",
       "      <td>2.452855</td>\n",
       "      <td>2.657462</td>\n",
       "      <td>2.480106</td>\n",
       "      <td>2.811486</td>\n",
       "      <td>3.098742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>138</td>\n",
       "      <td>tumoral</td>\n",
       "      <td>8.415823</td>\n",
       "      <td>4.579540</td>\n",
       "      <td>5.562391</td>\n",
       "      <td>6.185697</td>\n",
       "      <td>2.747484</td>\n",
       "      <td>6.961582</td>\n",
       "      <td>6.809795</td>\n",
       "      <td>7.158062</td>\n",
       "      <td>...</td>\n",
       "      <td>13.833406</td>\n",
       "      <td>13.853554</td>\n",
       "      <td>14.929317</td>\n",
       "      <td>14.749283</td>\n",
       "      <td>3.960969</td>\n",
       "      <td>2.593110</td>\n",
       "      <td>2.839218</td>\n",
       "      <td>2.483863</td>\n",
       "      <td>2.939876</td>\n",
       "      <td>2.761388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>139</td>\n",
       "      <td>tumoral</td>\n",
       "      <td>8.943121</td>\n",
       "      <td>6.577537</td>\n",
       "      <td>5.850425</td>\n",
       "      <td>5.846352</td>\n",
       "      <td>2.429107</td>\n",
       "      <td>6.492797</td>\n",
       "      <td>7.594337</td>\n",
       "      <td>6.060994</td>\n",
       "      <td>...</td>\n",
       "      <td>13.601130</td>\n",
       "      <td>13.604113</td>\n",
       "      <td>14.927910</td>\n",
       "      <td>14.753407</td>\n",
       "      <td>3.821295</td>\n",
       "      <td>2.629036</td>\n",
       "      <td>2.798198</td>\n",
       "      <td>2.487149</td>\n",
       "      <td>2.652175</td>\n",
       "      <td>2.600322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>141</td>\n",
       "      <td>tumoral</td>\n",
       "      <td>8.658734</td>\n",
       "      <td>4.624722</td>\n",
       "      <td>6.882949</td>\n",
       "      <td>5.880421</td>\n",
       "      <td>2.517857</td>\n",
       "      <td>7.519646</td>\n",
       "      <td>6.840436</td>\n",
       "      <td>6.290670</td>\n",
       "      <td>...</td>\n",
       "      <td>13.961861</td>\n",
       "      <td>13.933318</td>\n",
       "      <td>14.994563</td>\n",
       "      <td>14.803339</td>\n",
       "      <td>3.990229</td>\n",
       "      <td>2.862215</td>\n",
       "      <td>2.766505</td>\n",
       "      <td>2.229823</td>\n",
       "      <td>2.630706</td>\n",
       "      <td>2.588111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 54677 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   samples     type  1007_s_at   1053_at    117_at    121_at  1255_g_at  \\\n",
       "0      136  tumoral   8.767514  4.092134  6.006339  6.091951   2.738641   \n",
       "1      137  tumoral   9.177852  5.688045  5.599253  5.943833   2.667376   \n",
       "2      138  tumoral   8.415823  4.579540  5.562391  6.185697   2.747484   \n",
       "3      139  tumoral   8.943121  6.577537  5.850425  5.846352   2.429107   \n",
       "4      141  tumoral   8.658734  4.624722  6.882949  5.880421   2.517857   \n",
       "\n",
       "    1294_at   1316_at   1320_at  ...  AFFX-r2-Ec-bioD-3_at  \\\n",
       "0  7.707960  6.702969  5.203349  ...             13.852908   \n",
       "1  6.737484  7.336429  6.246692  ...             13.784316   \n",
       "2  6.961582  6.809795  7.158062  ...             13.833406   \n",
       "3  6.492797  7.594337  6.060994  ...             13.601130   \n",
       "4  7.519646  6.840436  6.290670  ...             13.961861   \n",
       "\n",
       "   AFFX-r2-Ec-bioD-5_at  AFFX-r2-P1-cre-3_at  AFFX-r2-P1-cre-5_at  \\\n",
       "0             13.854510            14.953006            14.824910   \n",
       "1             13.776588            14.944678            14.798245   \n",
       "2             13.853554            14.929317            14.749283   \n",
       "3             13.604113            14.927910            14.753407   \n",
       "4             13.933318            14.994563            14.803339   \n",
       "\n",
       "   AFFX-ThrX-3_at  AFFX-ThrX-5_at  AFFX-ThrX-M_at  AFFX-TrpnX-3_at  \\\n",
       "0        3.710845        2.493435        2.913242         2.360155   \n",
       "1        3.840439        2.452855        2.657462         2.480106   \n",
       "2        3.960969        2.593110        2.839218         2.483863   \n",
       "3        3.821295        2.629036        2.798198         2.487149   \n",
       "4        3.990229        2.862215        2.766505         2.229823   \n",
       "\n",
       "   AFFX-TrpnX-5_at  AFFX-TrpnX-M_at  \n",
       "0         2.758725         2.779035  \n",
       "1         2.811486         3.098742  \n",
       "2         2.939876         2.761388  \n",
       "3         2.652175         2.600322  \n",
       "4         2.630706         2.588111  \n",
       "\n",
       "[5 rows x 54677 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "main = pd.read_csv('../Raw_Data/Prostate_GSE46602.csv')\n",
    "main.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samples</th>\n",
       "      <th>type</th>\n",
       "      <th>1007_s_at</th>\n",
       "      <th>1053_at</th>\n",
       "      <th>117_at</th>\n",
       "      <th>121_at</th>\n",
       "      <th>1255_g_at</th>\n",
       "      <th>1294_at</th>\n",
       "      <th>1316_at</th>\n",
       "      <th>1320_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>136</td>\n",
       "      <td>tumoral</td>\n",
       "      <td>8.767514</td>\n",
       "      <td>4.092134</td>\n",
       "      <td>6.006339</td>\n",
       "      <td>6.091951</td>\n",
       "      <td>2.738641</td>\n",
       "      <td>7.707960</td>\n",
       "      <td>6.702969</td>\n",
       "      <td>5.203349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>137</td>\n",
       "      <td>tumoral</td>\n",
       "      <td>9.177852</td>\n",
       "      <td>5.688045</td>\n",
       "      <td>5.599253</td>\n",
       "      <td>5.943833</td>\n",
       "      <td>2.667376</td>\n",
       "      <td>6.737484</td>\n",
       "      <td>7.336429</td>\n",
       "      <td>6.246692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>138</td>\n",
       "      <td>tumoral</td>\n",
       "      <td>8.415823</td>\n",
       "      <td>4.579540</td>\n",
       "      <td>5.562391</td>\n",
       "      <td>6.185697</td>\n",
       "      <td>2.747484</td>\n",
       "      <td>6.961582</td>\n",
       "      <td>6.809795</td>\n",
       "      <td>7.158062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>139</td>\n",
       "      <td>tumoral</td>\n",
       "      <td>8.943121</td>\n",
       "      <td>6.577537</td>\n",
       "      <td>5.850425</td>\n",
       "      <td>5.846352</td>\n",
       "      <td>2.429107</td>\n",
       "      <td>6.492797</td>\n",
       "      <td>7.594337</td>\n",
       "      <td>6.060994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>141</td>\n",
       "      <td>tumoral</td>\n",
       "      <td>8.658734</td>\n",
       "      <td>4.624722</td>\n",
       "      <td>6.882949</td>\n",
       "      <td>5.880421</td>\n",
       "      <td>2.517857</td>\n",
       "      <td>7.519646</td>\n",
       "      <td>6.840436</td>\n",
       "      <td>6.290670</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   samples     type  1007_s_at   1053_at    117_at    121_at  1255_g_at  \\\n",
       "0      136  tumoral   8.767514  4.092134  6.006339  6.091951   2.738641   \n",
       "1      137  tumoral   9.177852  5.688045  5.599253  5.943833   2.667376   \n",
       "2      138  tumoral   8.415823  4.579540  5.562391  6.185697   2.747484   \n",
       "3      139  tumoral   8.943121  6.577537  5.850425  5.846352   2.429107   \n",
       "4      141  tumoral   8.658734  4.624722  6.882949  5.880421   2.517857   \n",
       "\n",
       "    1294_at   1316_at   1320_at  \n",
       "0  7.707960  6.702969  5.203349  \n",
       "1  6.737484  7.336429  6.246692  \n",
       "2  6.961582  6.809795  7.158062  \n",
       "3  6.492797  7.594337  6.060994  \n",
       "4  7.519646  6.840436  6.290670  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lim_main = main[main.columns[:10]]\n",
    "lim_main.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting functions.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile functions.py\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scipy.stats as st\n",
    "from sklearn.model_selection import train_test_split\n",
    "from imblearn.under_sampling import RandomUnderSampler\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay\n",
    "from sklearn.preprocessing import QuantileTransformer\n",
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, cohen_kappa_score\n",
    "\n",
    "def DropSamples(df:pd.DataFrame):\n",
    "    if 'samples' in df.columns:\n",
    "        df = df.drop('samples', axis=1)\n",
    "    return df\n",
    "\n",
    "\n",
    "def calculate_difference(df:pd.DataFrame, tumor_column = 'type'):\n",
    "    \"\"\"\n",
    "    Calculate the mean difference between two groups in a DataFrame based on the specified 'tumor_column'.\n",
    "\n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): Input DataFrame containing data for comparison.\n",
    "    - tumor_column (str, optional): The column defining the groups. Default is 'type'.\n",
    "\n",
    "    Returns:\n",
    "    - pd.DataFrame: A new DataFrame containing the means of each column for each group, \n",
    "                    as well as the differences between the means of the two groups.\n",
    "    \"\"\"    \n",
    "    df2 = DropSamples(df).copy()\n",
    "\n",
    "    means_df = df2.groupby(tumor_column).agg('mean').reset_index()\n",
    "\n",
    "    # define Min_values Dict\n",
    "    min_values = {}\n",
    "    for col in means_df.columns[1:]:\n",
    "        min_values[col] = means_df[col][0] - means_df[col][1]\n",
    "\n",
    "    min_values_df = pd.DataFrame([min_values])\n",
    "    min_values_df[tumor_column] = 'difference'\n",
    "\n",
    "    # return new concatted dataframe\n",
    "    return pd.concat([means_df, min_values_df], axis=0)\n",
    "\n",
    "\n",
    "def plotSTDEV(df:pd.DataFrame, tumor_value = 'tumoral', tumor_column = 'type'):\n",
    "    \"\"\"\n",
    "    Plot the standard deviation (STDEV) distribution of a specified 'tumor_value' within a DataFrame,\n",
    "    grouped by the 'tumor_column'. The function generates a boxplot and a histogram for visualizing the distribution.\n",
    "\n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): Input DataFrame containing data for visualization.\n",
    "    - tumor_value (str, optional): The specific value within the 'tumor_column' for which the STDEV is plotted.\n",
    "                                  Default is 'tumoral'.\n",
    "    \"\"\"\n",
    "    df2 = DropSamples(df).copy()\n",
    "\n",
    "    sted_df = df2.groupby(tumor_column).std().reset_index()\n",
    "    sted_df_melt = sted_df[sted_df[tumor_column] == tumor_value].drop(tumor_column, axis=1).melt()\n",
    "\n",
    "    f, ax  = plt.subplots(2,1,figsize=(10,5), sharex=True)\n",
    "    ax[0] = sns.boxplot(sted_df_melt, ax=ax[0], orient='h')\n",
    "    ax[1] = sns.histplot(sted_df_melt, ax=ax[1])\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "    return sted_df_melt\n",
    "\n",
    "\n",
    "# make function to plot the thing that we talked about\n",
    "\n",
    "def getLowSTEDVList(df:pd.DataFrame, tumor_value = 'tumoral', tumor_column = 'type'):\n",
    "    \"\"\"\n",
    "    Get a list of variables with a standard deviation (STDEV) less than 1 for a specific tumor group in a DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): Input DataFrame containing data for analysis.\n",
    "    - tumor_value (str, optional): The specific tumor group to analyze. Default is 'tumoral'.\n",
    "    - tumor_column (str, optional): The column defining the tumor groups. Default is 'type'.\n",
    "\n",
    "    Returns:\n",
    "    - list: A list of variables with STDEV values less than 1 for the specified tumor group.\n",
    "    \"\"\"\n",
    "    df2 = DropSamples(df).copy()\n",
    "\n",
    "    sted_df = df2.groupby(tumor_column).std().reset_index()\n",
    "    sted_df_melt = sted_df[sted_df[tumor_column] == tumor_value].drop(tumor_column, axis=1).melt()\n",
    "\n",
    "    return sted_df_melt[sted_df_melt['value']<1]['variable'].to_list()\n",
    "\n",
    "\n",
    "def getListDifferentialGenes(df:pd.DataFrame, psig=0.05, tumor_value = 'tumoral', tumor_column = 'type'):\n",
    "    \"\"\"\n",
    "    Identify differentially expressed genes between a specified tumor group and the rest of the samples in a DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): Input DataFrame containing gene expression data.\n",
    "    - psig (float, optional): Significance threshold for identifying differentially expressed genes. Default is 0.05.\n",
    "    - tumor_value (str, optional): The specific tumor group for which differential expression is assessed. Default is 'tumoral'.\n",
    "    - tumor_column (str, optional): The column defining the tumor groups. Default is 'type'.\n",
    "\n",
    "    Returns:\n",
    "    - list: A list of gene names that are differentially expressed in the specified tumor group based on a t-test.\n",
    "    \"\"\"    \n",
    "    df2 = df.copy()\n",
    "\n",
    "    df_cancer = df2[df2[tumor_column]==tumor_value]\n",
    "    df_normal = df2[df2[tumor_column]!=tumor_value]\n",
    "\n",
    "    differential_expressed_genes = []\n",
    "    for col in df_cancer.columns[2:]:\n",
    "        t, pvalue= st.ttest_ind(df_cancer[col],df_normal[col], equal_var = False, alternative = 'two-sided')\n",
    "        if pvalue < psig:\n",
    "            differential_expressed_genes.append(str(col))\n",
    "\n",
    "    return differential_expressed_genes\n",
    "\n",
    "\n",
    "def find_correlated_genes(df:pd.DataFrame, threshold=0.95):\n",
    "    \"\"\"\n",
    "    Identify and analyze genes with high correlation coefficients in a gene expression DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): Input DataFrame containing gene expression data.\n",
    "    - threshold (float, optional): The correlation coefficient threshold for identifying highly correlated genes. Default is 0.95.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple: A tuple containing three elements:\n",
    "        1. List: Information about correlated genes, including column names, correlation values, and index names.\n",
    "        2. List: Names of genes to be excluded to avoid redundancy.\n",
    "        3. pd.DataFrame: The correlation matrix of the input DataFrame.\n",
    "    \"\"\"\n",
    "    df2 = DropSamples(df).copy()\n",
    "    \n",
    "    r_values = df2.corr()\n",
    "    \n",
    "    correlated_genes_list = []\n",
    "    # Create a mask for values above the threshold\n",
    "    mask = (r_values.to_numpy() > threshold) & (r_values.index.to_numpy() != r_values.columns.to_numpy()[:, None])\n",
    "\n",
    "    # Extract the column and index names where the mask is True\n",
    "    correlated_columns, correlated_rows = np.where(mask)\n",
    "\n",
    "    for col, index in zip(r_values.columns[correlated_columns], r_values.index[correlated_rows]):\n",
    "        value = r_values.at[index, col]\n",
    "        correlated_genes_list.append([col, value, index])\n",
    "\n",
    "    exclude_list = []\n",
    "    included_columns = []\n",
    "\n",
    "    for col in correlated_genes_list:\n",
    "        if col[0] not in included_columns:\n",
    "            exclude_list.append(col[0])\n",
    "            included_columns.append(col[2])\n",
    "\n",
    "    return correlated_genes_list, exclude_list, r_values\n",
    "\n",
    "def find_correlated_genes_optimized(corr: pd.DataFrame, threshold=0.95):\n",
    "    \"\"\"\n",
    "    Identify and analyze genes with high correlation coefficients in a gene expression DataFrame.\n",
    "    Experimental function, \"optimized\" from the previous. This function loops over till no mutlicollinearity is found, but will take slightly longer.\n",
    "\n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): Input DataFrame containing gene expression data.\n",
    "    - threshold (float, optional): The correlation coefficient threshold for identifying highly correlated genes. Default is 0.95.\n",
    "\n",
    "    Returns:\n",
    "    - List: Names of genes to be excluded to avoid redundancy.\n",
    "    \"\"\"\n",
    "    exclude_list = []\n",
    "    corr2 = corr.copy()\n",
    "\n",
    "    # Create mask over all columns\n",
    "    while True:\n",
    "        mask = (corr2.to_numpy() > threshold) & (corr2.index.to_numpy() != corr2.columns.to_numpy()[:, None])\n",
    "        correlated_columns, correlated_rows = np.where(mask)\n",
    "    \n",
    "        correlated_genes_list = {}\n",
    "        for col, index in zip(corr2.columns[correlated_columns], corr2.index[correlated_rows]):\n",
    "            count = correlated_genes_list.get(col, 0)\n",
    "            count += 1\n",
    "            correlated_genes_list[col] = count\n",
    "\n",
    "        count_df = pd.DataFrame(correlated_genes_list, index=[0]).melt().sort_values('value', ascending=False)\n",
    "        highest_col_value = count_df['value'].iloc[0]\n",
    "\n",
    "        if highest_col_value <= 1:\n",
    "            break\n",
    "\n",
    "        highest_count_genes = count_df[count_df['value'] == highest_col_value]['variable'].tolist()\n",
    "\n",
    "        exclude_list.extend(highest_count_genes)\n",
    "        corr2 = corr2.drop(highest_count_genes, axis=1)\n",
    "        corr2 = corr2.drop(highest_count_genes)\n",
    "\n",
    "        print(f\"Loop completed, highest value was {highest_col_value} from genes {highest_count_genes}\")\n",
    "\n",
    "    return exclude_list\n",
    "\n",
    "def trainLogisticModel(df: pd.DataFrame, ColumnToPredict='type', tumor_value = 'tumoral'):\n",
    "    \"\"\"\n",
    "    Train a logistic regression model to predict a specified column in a DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): Input DataFrame containing data for training and testing the model.\n",
    "    - ColumnToPredict (str, optional): The column to be predicted using logistic regression. Default is 'type'.\n",
    "    - tumor_value (str, optional): The specific tumor group to be treated as the positive label. Default is 'tumoral'.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple: A tuple containing three elements:\n",
    "        1. LogisticRegression: Trained logistic regression model.\n",
    "        2. pd.DataFrame: Confusion matrix for model evaluation.\n",
    "        3. dict: Dictionary containing various classification scores (e.g., accuracy, precision, recall).\n",
    "    \"\"\"\n",
    "    df2 = DropSamples(df).copy()\n",
    "\n",
    "    # Split into X/Y based on ColumnToPredict\n",
    "    y = df2[ColumnToPredict]\n",
    "    X = df2.drop([ColumnToPredict], axis=1)\n",
    "    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=1337)\n",
    "\n",
    "    # Peform underSampler if values are not equal\n",
    "    if df2[ColumnToPredict].value_counts()[0] != df2[ColumnToPredict].value_counts()[1]:\n",
    "        rus = RandomUnderSampler(random_state=69)\n",
    "        X_train, y_train = rus.fit_resample(X_train, y_train)\n",
    "\n",
    "    # Create LogisticRegression model\n",
    "    classification = LogisticRegression(random_state=0, solver='lbfgs')\n",
    "    classification.fit(X_train, y_train)\n",
    "\n",
    "    # Create Prediction values\n",
    "    y_pred = classification.predict(X=X_test)\n",
    "\n",
    "    # return y_pred, y_test\n",
    "    model, cm, scores = LogisticModelView(model=classification, y_pred=y_pred, y_test=y_test.values, pos_label=tumor_value)\n",
    "    return model, cm, scores\n",
    "\n",
    "def predict(model: LogisticRegression, X, tumor_value = 'tumoral'):\n",
    "    \"\"\"\n",
    "    Do not use, not ready yet. Maybe will never be ready tbh might not really need this one :)\n",
    "    \"\"\"\n",
    "    model, cm, scores = LogisticModelView(model=model, y_pred=model.predict(X=X), y_test=X, pos_label=tumor_value)\n",
    "    return model, cm, scores\n",
    "\n",
    "def LogisticModelView(model: LogisticRegression, y_pred, y_test, pos_label):\n",
    "    \"\"\"\n",
    "    Evaluate and visualize the performance of a logistic regression model using a confusion matrix.\n",
    "\n",
    "    Parameters:\n",
    "    - model (LogisticRegression): Trained logistic regression model.\n",
    "    - y_pred: Predicted values from the model.\n",
    "    - y_test: True labels for evaluation.\n",
    "    - pos_label (str): The positive label, typically the specific tumor group.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple: A tuple containing three elements:\n",
    "        1. LogisticRegression: Trained logistic regression model.\n",
    "        2. pd.DataFrame: Confusion matrix for model evaluation.\n",
    "        3. dict: Dictionary containing various classification scores (e.g., accuracy, precision, recall).\n",
    "    \"\"\"\n",
    "    # Create Confusion Matrix\n",
    "    cm = confusion_matrix(y_test, y_pred)\n",
    "\n",
    "    disp = ConfusionMatrixDisplay(cm, display_labels=model.classes_)\n",
    "    disp.plot()\n",
    "\n",
    "    # Print Precision of CM\n",
    "    scores = {\n",
    "        'accuracy': accuracy_score(y_test, y_pred),\n",
    "        'precision': precision_score(y_true=y_test, y_pred=y_pred, pos_label=pos_label),\n",
    "        'recall': recall_score(y_test, y_pred, pos_label=pos_label),\n",
    "        'f1': f1_score(y_test, y_pred, pos_label=pos_label),\n",
    "        'kappa': cohen_kappa_score(y_test, y_pred)\n",
    "    }\n",
    "\n",
    "    return model, cm, scores\n",
    "\n",
    "\n",
    "\n",
    "def quantileTransformer(df:pd.DataFrame, qt = None):\n",
    "    \"\"\"\n",
    "    Apply quantile transformation to a DataFrame, transforming numerical columns to follow a normal distribution.\n",
    "\n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): Input DataFrame containing numerical data for transformation.\n",
    "    - qt (QuantileTransformer or None, optional): An existing QuantileTransformer object to apply, or None to fit a new one. Default is None.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple: A tuple containing two elements:\n",
    "        1. pd.DataFrame: Transformed DataFrame with numerical columns following a normal distribution.\n",
    "        2. QuantileTransformer: The fitted or provided QuantileTransformer object.\n",
    "    \"\"\"\n",
    "    df_transformed = df.copy()\n",
    "\n",
    "    if qt is None:\n",
    "        # Fit new dataset, if none is given\n",
    "        qt = QuantileTransformer (output_distribution ='normal')\n",
    "        qt.fit(df_transformed[df_transformed.columns[2:]])\n",
    "    \n",
    "    # Apply Transformation to data set\n",
    "    df_transformed[df_transformed.columns[2:]] = qt.transform(df_transformed[df_transformed.columns[2:]])\n",
    "    return df_transformed, qt\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "lst = getListDifferentialGenes(main)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_main = main[[\n",
    "    'samples', 'type'\n",
    "] + lst]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "corr = new_main.drop(['samples', 'type'], axis=1).corr()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loop completed, highest value was 9 from genes ['AFFX-BioC-5_at', 'AFFX-BioB-3_at', 'AFFX-r2-Ec-bioC-3_at', 'AFFX-r2-Ec-bioB-M_at', 'AFFX-BioB-M_at', 'AFFX-BioC-3_at']\n",
      "Loop completed, highest value was 7 from genes ['211456_x_at']\n",
      "Loop completed, highest value was 6 from genes ['207730_x_at']\n",
      "Loop completed, highest value was 5 from genes ['211417_x_at', '211983_x_at', '204745_x_at', '206461_x_at', '207131_x_at', '208284_x_at', '224585_x_at', '211416_x_at', '212988_x_at']\n",
      "Loop completed, highest value was 4 from genes ['211765_x_at', '211978_x_at', '209863_s_at', '212661_x_at', '201293_x_at', '211378_x_at']\n",
      "Loop completed, highest value was 3 from genes ['208929_x_at', '217398_x_at', '213453_x_at', '209813_x_at', '241926_s_at', '1563392_at', '210427_x_at', '212734_x_at', '212581_x_at', '204538_x_at', '210679_x_at', '229447_x_at', '211025_x_at', '213541_s_at', '215806_x_at', '211144_x_at', '212191_x_at', '211527_x_at', '203242_s_at', '216920_s_at', '201590_x_at', '211940_x_at', '202343_x_at', '211626_x_at', '213503_x_at', '214870_x_at', '208755_x_at', '213735_s_at', '213736_at', '214351_x_at', '208816_x_at', '213826_s_at', '213828_x_at', '208527_x_at']\n",
      "Loop completed, highest value was 2 from genes ['AFFX-r2-Ec-bioD-5_at', '209366_x_at', '201154_x_at', '204418_x_at', '201033_x_at', '204532_x_at', '204550_x_at', '211710_x_at', '200809_x_at', '211737_x_at', '226835_s_at', '215726_s_at', '200089_s_at', '210183_x_at', '211972_x_at', '226227_x_at', '215333_x_at', '211487_x_at', '201255_x_at', '215108_x_at', '216623_x_at', '202649_x_at', '202648_at', '217740_x_at', '217756_x_at', '202428_x_at', '232482_at', '210588_x_at', '209426_s_at', '217165_x_at', '209425_at', '209424_s_at', '234873_x_at', '209389_x_at', '236121_at', '211070_x_at', '215125_s_at', '200088_x_at', '208768_x_at', '206094_x_at', 'AFFX-r2-Ec-bioD-3_at', 'AFFX-r2-Ec-bioB-5_at', 'AFFX-r2-Ec-bioB-3_at', 'AFFX-BioDn-3_at', 'AFFX-BioB-5_at', '208581_x_at', '207843_x_at', '221775_x_at', '213414_s_at', '207127_s_at', '224625_x_at', '214271_x_at', '221424_s_at', '208856_x_at', '210110_x_at', '225876_at', '224930_x_at', '224915_x_at', '212433_x_at', '225875_s_at', '220960_x_at', '214774_x_at']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['AFFX-BioC-5_at',\n",
       " 'AFFX-BioB-3_at',\n",
       " 'AFFX-r2-Ec-bioC-3_at',\n",
       " 'AFFX-r2-Ec-bioB-M_at',\n",
       " 'AFFX-BioB-M_at',\n",
       " 'AFFX-BioC-3_at',\n",
       " '211456_x_at',\n",
       " '207730_x_at',\n",
       " '211417_x_at',\n",
       " '211983_x_at',\n",
       " '204745_x_at',\n",
       " '206461_x_at',\n",
       " '207131_x_at',\n",
       " '208284_x_at',\n",
       " '224585_x_at',\n",
       " '211416_x_at',\n",
       " '212988_x_at',\n",
       " '211765_x_at',\n",
       " '211978_x_at',\n",
       " '209863_s_at',\n",
       " '212661_x_at',\n",
       " '201293_x_at',\n",
       " '211378_x_at',\n",
       " '208929_x_at',\n",
       " '217398_x_at',\n",
       " '213453_x_at',\n",
       " '209813_x_at',\n",
       " '241926_s_at',\n",
       " '1563392_at',\n",
       " '210427_x_at',\n",
       " '212734_x_at',\n",
       " '212581_x_at',\n",
       " '204538_x_at',\n",
       " '210679_x_at',\n",
       " '229447_x_at',\n",
       " '211025_x_at',\n",
       " '213541_s_at',\n",
       " '215806_x_at',\n",
       " '211144_x_at',\n",
       " '212191_x_at',\n",
       " '211527_x_at',\n",
       " '203242_s_at',\n",
       " '216920_s_at',\n",
       " '201590_x_at',\n",
       " '211940_x_at',\n",
       " '202343_x_at',\n",
       " '211626_x_at',\n",
       " '213503_x_at',\n",
       " '214870_x_at',\n",
       " '208755_x_at',\n",
       " '213735_s_at',\n",
       " '213736_at',\n",
       " '214351_x_at',\n",
       " '208816_x_at',\n",
       " '213826_s_at',\n",
       " '213828_x_at',\n",
       " '208527_x_at',\n",
       " 'AFFX-r2-Ec-bioD-5_at',\n",
       " '209366_x_at',\n",
       " '201154_x_at',\n",
       " '204418_x_at',\n",
       " '201033_x_at',\n",
       " '204532_x_at',\n",
       " '204550_x_at',\n",
       " '211710_x_at',\n",
       " '200809_x_at',\n",
       " '211737_x_at',\n",
       " '226835_s_at',\n",
       " '215726_s_at',\n",
       " '200089_s_at',\n",
       " '210183_x_at',\n",
       " '211972_x_at',\n",
       " '226227_x_at',\n",
       " '215333_x_at',\n",
       " '211487_x_at',\n",
       " '201255_x_at',\n",
       " '215108_x_at',\n",
       " '216623_x_at',\n",
       " '202649_x_at',\n",
       " '202648_at',\n",
       " '217740_x_at',\n",
       " '217756_x_at',\n",
       " '202428_x_at',\n",
       " '232482_at',\n",
       " '210588_x_at',\n",
       " '209426_s_at',\n",
       " '217165_x_at',\n",
       " '209425_at',\n",
       " '209424_s_at',\n",
       " '234873_x_at',\n",
       " '209389_x_at',\n",
       " '236121_at',\n",
       " '211070_x_at',\n",
       " '215125_s_at',\n",
       " '200088_x_at',\n",
       " '208768_x_at',\n",
       " '206094_x_at',\n",
       " 'AFFX-r2-Ec-bioD-3_at',\n",
       " 'AFFX-r2-Ec-bioB-5_at',\n",
       " 'AFFX-r2-Ec-bioB-3_at',\n",
       " 'AFFX-BioDn-3_at',\n",
       " 'AFFX-BioB-5_at',\n",
       " '208581_x_at',\n",
       " '207843_x_at',\n",
       " '221775_x_at',\n",
       " '213414_s_at',\n",
       " '207127_s_at',\n",
       " '224625_x_at',\n",
       " '214271_x_at',\n",
       " '221424_s_at',\n",
       " '208856_x_at',\n",
       " '210110_x_at',\n",
       " '225876_at',\n",
       " '224930_x_at',\n",
       " '224915_x_at',\n",
       " '212433_x_at',\n",
       " '225875_s_at',\n",
       " '220960_x_at',\n",
       " '214774_x_at']"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b = find_correlated_genes_optimized(corr, 0.95)\n",
    "b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "119\n"
     ]
    }
   ],
   "source": [
    "print(len(b))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "b.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1294_at</th>\n",
       "      <th>1316_at</th>\n",
       "      <th>1431_at</th>\n",
       "      <th>1438_at</th>\n",
       "      <th>1487_at</th>\n",
       "      <th>1552256_a_at</th>\n",
       "      <th>1552257_a_at</th>\n",
       "      <th>1552258_at</th>\n",
       "      <th>1552263_at</th>\n",
       "      <th>1552264_a_at</th>\n",
       "      <th>...</th>\n",
       "      <th>AFFX-r2-Ec-bioB-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioB-5_at</th>\n",
       "      <th>AFFX-r2-Ec-bioB-M_at</th>\n",
       "      <th>AFFX-r2-Ec-bioC-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioC-5_at</th>\n",
       "      <th>AFFX-r2-Ec-bioD-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioD-5_at</th>\n",
       "      <th>AFFX-r2-P1-cre-3_at</th>\n",
       "      <th>AFFX-r2-P1-cre-5_at</th>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1294_at</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.196297</td>\n",
       "      <td>0.265514</td>\n",
       "      <td>-0.158850</td>\n",
       "      <td>-0.390742</td>\n",
       "      <td>-0.322735</td>\n",
       "      <td>-0.629106</td>\n",
       "      <td>0.619377</td>\n",
       "      <td>-0.118182</td>\n",
       "      <td>-0.243256</td>\n",
       "      <td>...</td>\n",
       "      <td>0.639111</td>\n",
       "      <td>0.641003</td>\n",
       "      <td>0.665970</td>\n",
       "      <td>0.673197</td>\n",
       "      <td>0.687415</td>\n",
       "      <td>0.506664</td>\n",
       "      <td>0.513898</td>\n",
       "      <td>0.140528</td>\n",
       "      <td>0.148749</td>\n",
       "      <td>0.037070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1316_at</th>\n",
       "      <td>-0.196297</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.107147</td>\n",
       "      <td>0.297960</td>\n",
       "      <td>0.401575</td>\n",
       "      <td>0.437606</td>\n",
       "      <td>0.492194</td>\n",
       "      <td>-0.238358</td>\n",
       "      <td>0.287821</td>\n",
       "      <td>0.222830</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.044619</td>\n",
       "      <td>-0.012882</td>\n",
       "      <td>-0.056788</td>\n",
       "      <td>-0.128206</td>\n",
       "      <td>-0.148492</td>\n",
       "      <td>-0.342041</td>\n",
       "      <td>-0.267291</td>\n",
       "      <td>-0.502214</td>\n",
       "      <td>-0.525167</td>\n",
       "      <td>0.257226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1431_at</th>\n",
       "      <td>0.265514</td>\n",
       "      <td>-0.107147</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.026961</td>\n",
       "      <td>-0.093128</td>\n",
       "      <td>-0.239341</td>\n",
       "      <td>-0.246241</td>\n",
       "      <td>0.172745</td>\n",
       "      <td>-0.272039</td>\n",
       "      <td>-0.196795</td>\n",
       "      <td>...</td>\n",
       "      <td>0.344679</td>\n",
       "      <td>0.318784</td>\n",
       "      <td>0.317479</td>\n",
       "      <td>0.332442</td>\n",
       "      <td>0.277273</td>\n",
       "      <td>0.143749</td>\n",
       "      <td>0.138784</td>\n",
       "      <td>-0.163597</td>\n",
       "      <td>-0.168589</td>\n",
       "      <td>0.065501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438_at</th>\n",
       "      <td>-0.158850</td>\n",
       "      <td>0.297960</td>\n",
       "      <td>-0.026961</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.444146</td>\n",
       "      <td>0.296720</td>\n",
       "      <td>0.441981</td>\n",
       "      <td>-0.112081</td>\n",
       "      <td>0.111677</td>\n",
       "      <td>-0.172533</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.219205</td>\n",
       "      <td>-0.196298</td>\n",
       "      <td>-0.257506</td>\n",
       "      <td>-0.360003</td>\n",
       "      <td>-0.399195</td>\n",
       "      <td>-0.477992</td>\n",
       "      <td>-0.463958</td>\n",
       "      <td>-0.283587</td>\n",
       "      <td>-0.551300</td>\n",
       "      <td>0.612591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1487_at</th>\n",
       "      <td>-0.390742</td>\n",
       "      <td>0.401575</td>\n",
       "      <td>-0.093128</td>\n",
       "      <td>0.444146</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.395359</td>\n",
       "      <td>0.616744</td>\n",
       "      <td>-0.144776</td>\n",
       "      <td>0.321502</td>\n",
       "      <td>0.183275</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.335935</td>\n",
       "      <td>-0.332841</td>\n",
       "      <td>-0.372503</td>\n",
       "      <td>-0.476981</td>\n",
       "      <td>-0.495966</td>\n",
       "      <td>-0.568344</td>\n",
       "      <td>-0.477054</td>\n",
       "      <td>-0.389978</td>\n",
       "      <td>-0.528519</td>\n",
       "      <td>0.256257</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 19094 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          1294_at   1316_at   1431_at   1438_at   1487_at  1552256_a_at  \\\n",
       "1294_at  1.000000 -0.196297  0.265514 -0.158850 -0.390742     -0.322735   \n",
       "1316_at -0.196297  1.000000 -0.107147  0.297960  0.401575      0.437606   \n",
       "1431_at  0.265514 -0.107147  1.000000 -0.026961 -0.093128     -0.239341   \n",
       "1438_at -0.158850  0.297960 -0.026961  1.000000  0.444146      0.296720   \n",
       "1487_at -0.390742  0.401575 -0.093128  0.444146  1.000000      0.395359   \n",
       "\n",
       "         1552257_a_at  1552258_at  1552263_at  1552264_a_at  ...  \\\n",
       "1294_at     -0.629106    0.619377   -0.118182     -0.243256  ...   \n",
       "1316_at      0.492194   -0.238358    0.287821      0.222830  ...   \n",
       "1431_at     -0.246241    0.172745   -0.272039     -0.196795  ...   \n",
       "1438_at      0.441981   -0.112081    0.111677     -0.172533  ...   \n",
       "1487_at      0.616744   -0.144776    0.321502      0.183275  ...   \n",
       "\n",
       "         AFFX-r2-Ec-bioB-3_at  AFFX-r2-Ec-bioB-5_at  AFFX-r2-Ec-bioB-M_at  \\\n",
       "1294_at              0.639111              0.641003              0.665970   \n",
       "1316_at             -0.044619             -0.012882             -0.056788   \n",
       "1431_at              0.344679              0.318784              0.317479   \n",
       "1438_at             -0.219205             -0.196298             -0.257506   \n",
       "1487_at             -0.335935             -0.332841             -0.372503   \n",
       "\n",
       "         AFFX-r2-Ec-bioC-3_at  AFFX-r2-Ec-bioC-5_at  AFFX-r2-Ec-bioD-3_at  \\\n",
       "1294_at              0.673197              0.687415              0.506664   \n",
       "1316_at             -0.128206             -0.148492             -0.342041   \n",
       "1431_at              0.332442              0.277273              0.143749   \n",
       "1438_at             -0.360003             -0.399195             -0.477992   \n",
       "1487_at             -0.476981             -0.495966             -0.568344   \n",
       "\n",
       "         AFFX-r2-Ec-bioD-5_at  AFFX-r2-P1-cre-3_at  AFFX-r2-P1-cre-5_at  \\\n",
       "1294_at              0.513898             0.140528             0.148749   \n",
       "1316_at             -0.267291            -0.502214            -0.525167   \n",
       "1431_at              0.138784            -0.163597            -0.168589   \n",
       "1438_at             -0.463958            -0.283587            -0.551300   \n",
       "1487_at             -0.477054            -0.389978            -0.528519   \n",
       "\n",
       "         AFFX-TrpnX-M_at  \n",
       "1294_at         0.037070  \n",
       "1316_at         0.257226  \n",
       "1431_at         0.065501  \n",
       "1438_at         0.612591  \n",
       "1487_at         0.256257  \n",
       "\n",
       "[5 rows x 19094 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "corr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = corr.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1294_at</th>\n",
       "      <th>1316_at</th>\n",
       "      <th>1431_at</th>\n",
       "      <th>1438_at</th>\n",
       "      <th>1487_at</th>\n",
       "      <th>1552256_a_at</th>\n",
       "      <th>1552257_a_at</th>\n",
       "      <th>1552258_at</th>\n",
       "      <th>1552263_at</th>\n",
       "      <th>1552264_a_at</th>\n",
       "      <th>...</th>\n",
       "      <th>AFFX-r2-Ec-bioB-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioB-5_at</th>\n",
       "      <th>AFFX-r2-Ec-bioB-M_at</th>\n",
       "      <th>AFFX-r2-Ec-bioC-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioC-5_at</th>\n",
       "      <th>AFFX-r2-Ec-bioD-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioD-5_at</th>\n",
       "      <th>AFFX-r2-P1-cre-3_at</th>\n",
       "      <th>AFFX-r2-P1-cre-5_at</th>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1294_at</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.196297</td>\n",
       "      <td>0.265514</td>\n",
       "      <td>-0.158850</td>\n",
       "      <td>-0.390742</td>\n",
       "      <td>-0.322735</td>\n",
       "      <td>-0.629106</td>\n",
       "      <td>0.619377</td>\n",
       "      <td>-0.118182</td>\n",
       "      <td>-0.243256</td>\n",
       "      <td>...</td>\n",
       "      <td>0.639111</td>\n",
       "      <td>0.641003</td>\n",
       "      <td>0.665970</td>\n",
       "      <td>0.673197</td>\n",
       "      <td>0.687415</td>\n",
       "      <td>0.506664</td>\n",
       "      <td>0.513898</td>\n",
       "      <td>0.140528</td>\n",
       "      <td>0.148749</td>\n",
       "      <td>0.037070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1316_at</th>\n",
       "      <td>-0.196297</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.107147</td>\n",
       "      <td>0.297960</td>\n",
       "      <td>0.401575</td>\n",
       "      <td>0.437606</td>\n",
       "      <td>0.492194</td>\n",
       "      <td>-0.238358</td>\n",
       "      <td>0.287821</td>\n",
       "      <td>0.222830</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.044619</td>\n",
       "      <td>-0.012882</td>\n",
       "      <td>-0.056788</td>\n",
       "      <td>-0.128206</td>\n",
       "      <td>-0.148492</td>\n",
       "      <td>-0.342041</td>\n",
       "      <td>-0.267291</td>\n",
       "      <td>-0.502214</td>\n",
       "      <td>-0.525167</td>\n",
       "      <td>0.257226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1431_at</th>\n",
       "      <td>0.265514</td>\n",
       "      <td>-0.107147</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.026961</td>\n",
       "      <td>-0.093128</td>\n",
       "      <td>-0.239341</td>\n",
       "      <td>-0.246241</td>\n",
       "      <td>0.172745</td>\n",
       "      <td>-0.272039</td>\n",
       "      <td>-0.196795</td>\n",
       "      <td>...</td>\n",
       "      <td>0.344679</td>\n",
       "      <td>0.318784</td>\n",
       "      <td>0.317479</td>\n",
       "      <td>0.332442</td>\n",
       "      <td>0.277273</td>\n",
       "      <td>0.143749</td>\n",
       "      <td>0.138784</td>\n",
       "      <td>-0.163597</td>\n",
       "      <td>-0.168589</td>\n",
       "      <td>0.065501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438_at</th>\n",
       "      <td>-0.158850</td>\n",
       "      <td>0.297960</td>\n",
       "      <td>-0.026961</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.444146</td>\n",
       "      <td>0.296720</td>\n",
       "      <td>0.441981</td>\n",
       "      <td>-0.112081</td>\n",
       "      <td>0.111677</td>\n",
       "      <td>-0.172533</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.219205</td>\n",
       "      <td>-0.196298</td>\n",
       "      <td>-0.257506</td>\n",
       "      <td>-0.360003</td>\n",
       "      <td>-0.399195</td>\n",
       "      <td>-0.477992</td>\n",
       "      <td>-0.463958</td>\n",
       "      <td>-0.283587</td>\n",
       "      <td>-0.551300</td>\n",
       "      <td>0.612591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1487_at</th>\n",
       "      <td>-0.390742</td>\n",
       "      <td>0.401575</td>\n",
       "      <td>-0.093128</td>\n",
       "      <td>0.444146</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.395359</td>\n",
       "      <td>0.616744</td>\n",
       "      <td>-0.144776</td>\n",
       "      <td>0.321502</td>\n",
       "      <td>0.183275</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.335935</td>\n",
       "      <td>-0.332841</td>\n",
       "      <td>-0.372503</td>\n",
       "      <td>-0.476981</td>\n",
       "      <td>-0.495966</td>\n",
       "      <td>-0.568344</td>\n",
       "      <td>-0.477054</td>\n",
       "      <td>-0.389978</td>\n",
       "      <td>-0.528519</td>\n",
       "      <td>0.256257</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 19094 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          1294_at   1316_at   1431_at   1438_at   1487_at  1552256_a_at  \\\n",
       "1294_at  1.000000 -0.196297  0.265514 -0.158850 -0.390742     -0.322735   \n",
       "1316_at -0.196297  1.000000 -0.107147  0.297960  0.401575      0.437606   \n",
       "1431_at  0.265514 -0.107147  1.000000 -0.026961 -0.093128     -0.239341   \n",
       "1438_at -0.158850  0.297960 -0.026961  1.000000  0.444146      0.296720   \n",
       "1487_at -0.390742  0.401575 -0.093128  0.444146  1.000000      0.395359   \n",
       "\n",
       "         1552257_a_at  1552258_at  1552263_at  1552264_a_at  ...  \\\n",
       "1294_at     -0.629106    0.619377   -0.118182     -0.243256  ...   \n",
       "1316_at      0.492194   -0.238358    0.287821      0.222830  ...   \n",
       "1431_at     -0.246241    0.172745   -0.272039     -0.196795  ...   \n",
       "1438_at      0.441981   -0.112081    0.111677     -0.172533  ...   \n",
       "1487_at      0.616744   -0.144776    0.321502      0.183275  ...   \n",
       "\n",
       "         AFFX-r2-Ec-bioB-3_at  AFFX-r2-Ec-bioB-5_at  AFFX-r2-Ec-bioB-M_at  \\\n",
       "1294_at              0.639111              0.641003              0.665970   \n",
       "1316_at             -0.044619             -0.012882             -0.056788   \n",
       "1431_at              0.344679              0.318784              0.317479   \n",
       "1438_at             -0.219205             -0.196298             -0.257506   \n",
       "1487_at             -0.335935             -0.332841             -0.372503   \n",
       "\n",
       "         AFFX-r2-Ec-bioC-3_at  AFFX-r2-Ec-bioC-5_at  AFFX-r2-Ec-bioD-3_at  \\\n",
       "1294_at              0.673197              0.687415              0.506664   \n",
       "1316_at             -0.128206             -0.148492             -0.342041   \n",
       "1431_at              0.332442              0.277273              0.143749   \n",
       "1438_at             -0.360003             -0.399195             -0.477992   \n",
       "1487_at             -0.476981             -0.495966             -0.568344   \n",
       "\n",
       "         AFFX-r2-Ec-bioD-5_at  AFFX-r2-P1-cre-3_at  AFFX-r2-P1-cre-5_at  \\\n",
       "1294_at              0.513898             0.140528             0.148749   \n",
       "1316_at             -0.267291            -0.502214            -0.525167   \n",
       "1431_at              0.138784            -0.163597            -0.168589   \n",
       "1438_at             -0.463958            -0.283587            -0.551300   \n",
       "1487_at             -0.477054            -0.389978            -0.528519   \n",
       "\n",
       "         AFFX-TrpnX-M_at  \n",
       "1294_at         0.037070  \n",
       "1316_at         0.257226  \n",
       "1431_at         0.065501  \n",
       "1438_at         0.612591  \n",
       "1487_at         0.256257  \n",
       "\n",
       "[5 rows x 19094 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1294_at</th>\n",
       "      <th>1316_at</th>\n",
       "      <th>1431_at</th>\n",
       "      <th>1438_at</th>\n",
       "      <th>1487_at</th>\n",
       "      <th>1552256_a_at</th>\n",
       "      <th>1552257_a_at</th>\n",
       "      <th>1552258_at</th>\n",
       "      <th>1552263_at</th>\n",
       "      <th>1552264_a_at</th>\n",
       "      <th>...</th>\n",
       "      <th>AFFX-r2-Ec-bioB-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioB-5_at</th>\n",
       "      <th>AFFX-r2-Ec-bioB-M_at</th>\n",
       "      <th>AFFX-r2-Ec-bioC-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioC-5_at</th>\n",
       "      <th>AFFX-r2-Ec-bioD-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioD-5_at</th>\n",
       "      <th>AFFX-r2-P1-cre-3_at</th>\n",
       "      <th>AFFX-r2-P1-cre-5_at</th>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1316_at</th>\n",
       "      <td>-0.196297</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.107147</td>\n",
       "      <td>0.297960</td>\n",
       "      <td>0.401575</td>\n",
       "      <td>0.437606</td>\n",
       "      <td>0.492194</td>\n",
       "      <td>-0.238358</td>\n",
       "      <td>0.287821</td>\n",
       "      <td>0.222830</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.044619</td>\n",
       "      <td>-0.012882</td>\n",
       "      <td>-0.056788</td>\n",
       "      <td>-0.128206</td>\n",
       "      <td>-0.148492</td>\n",
       "      <td>-0.342041</td>\n",
       "      <td>-0.267291</td>\n",
       "      <td>-0.502214</td>\n",
       "      <td>-0.525167</td>\n",
       "      <td>0.257226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1431_at</th>\n",
       "      <td>0.265514</td>\n",
       "      <td>-0.107147</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.026961</td>\n",
       "      <td>-0.093128</td>\n",
       "      <td>-0.239341</td>\n",
       "      <td>-0.246241</td>\n",
       "      <td>0.172745</td>\n",
       "      <td>-0.272039</td>\n",
       "      <td>-0.196795</td>\n",
       "      <td>...</td>\n",
       "      <td>0.344679</td>\n",
       "      <td>0.318784</td>\n",
       "      <td>0.317479</td>\n",
       "      <td>0.332442</td>\n",
       "      <td>0.277273</td>\n",
       "      <td>0.143749</td>\n",
       "      <td>0.138784</td>\n",
       "      <td>-0.163597</td>\n",
       "      <td>-0.168589</td>\n",
       "      <td>0.065501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438_at</th>\n",
       "      <td>-0.158850</td>\n",
       "      <td>0.297960</td>\n",
       "      <td>-0.026961</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.444146</td>\n",
       "      <td>0.296720</td>\n",
       "      <td>0.441981</td>\n",
       "      <td>-0.112081</td>\n",
       "      <td>0.111677</td>\n",
       "      <td>-0.172533</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.219205</td>\n",
       "      <td>-0.196298</td>\n",
       "      <td>-0.257506</td>\n",
       "      <td>-0.360003</td>\n",
       "      <td>-0.399195</td>\n",
       "      <td>-0.477992</td>\n",
       "      <td>-0.463958</td>\n",
       "      <td>-0.283587</td>\n",
       "      <td>-0.551300</td>\n",
       "      <td>0.612591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1487_at</th>\n",
       "      <td>-0.390742</td>\n",
       "      <td>0.401575</td>\n",
       "      <td>-0.093128</td>\n",
       "      <td>0.444146</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.395359</td>\n",
       "      <td>0.616744</td>\n",
       "      <td>-0.144776</td>\n",
       "      <td>0.321502</td>\n",
       "      <td>0.183275</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.335935</td>\n",
       "      <td>-0.332841</td>\n",
       "      <td>-0.372503</td>\n",
       "      <td>-0.476981</td>\n",
       "      <td>-0.495966</td>\n",
       "      <td>-0.568344</td>\n",
       "      <td>-0.477054</td>\n",
       "      <td>-0.389978</td>\n",
       "      <td>-0.528519</td>\n",
       "      <td>0.256257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1552256_a_at</th>\n",
       "      <td>-0.322735</td>\n",
       "      <td>0.437606</td>\n",
       "      <td>-0.239341</td>\n",
       "      <td>0.296720</td>\n",
       "      <td>0.395359</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.578383</td>\n",
       "      <td>-0.170316</td>\n",
       "      <td>0.220145</td>\n",
       "      <td>0.501943</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.275150</td>\n",
       "      <td>-0.208974</td>\n",
       "      <td>-0.255458</td>\n",
       "      <td>-0.359736</td>\n",
       "      <td>-0.365745</td>\n",
       "      <td>-0.434607</td>\n",
       "      <td>-0.359832</td>\n",
       "      <td>-0.262267</td>\n",
       "      <td>-0.424885</td>\n",
       "      <td>0.385145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-r2-Ec-bioD-3_at</th>\n",
       "      <td>0.506664</td>\n",
       "      <td>-0.342041</td>\n",
       "      <td>0.143749</td>\n",
       "      <td>-0.477992</td>\n",
       "      <td>-0.568344</td>\n",
       "      <td>-0.434607</td>\n",
       "      <td>-0.658133</td>\n",
       "      <td>0.429023</td>\n",
       "      <td>-0.184792</td>\n",
       "      <td>-0.123851</td>\n",
       "      <td>...</td>\n",
       "      <td>0.656057</td>\n",
       "      <td>0.632837</td>\n",
       "      <td>0.678121</td>\n",
       "      <td>0.757525</td>\n",
       "      <td>0.765996</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.973461</td>\n",
       "      <td>0.588087</td>\n",
       "      <td>0.728331</td>\n",
       "      <td>-0.456953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-r2-Ec-bioD-5_at</th>\n",
       "      <td>0.513898</td>\n",
       "      <td>-0.267291</td>\n",
       "      <td>0.138784</td>\n",
       "      <td>-0.463958</td>\n",
       "      <td>-0.477054</td>\n",
       "      <td>-0.359832</td>\n",
       "      <td>-0.594893</td>\n",
       "      <td>0.442233</td>\n",
       "      <td>-0.120095</td>\n",
       "      <td>-0.115855</td>\n",
       "      <td>...</td>\n",
       "      <td>0.669071</td>\n",
       "      <td>0.642518</td>\n",
       "      <td>0.684320</td>\n",
       "      <td>0.754380</td>\n",
       "      <td>0.762137</td>\n",
       "      <td>0.973461</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.531057</td>\n",
       "      <td>0.690192</td>\n",
       "      <td>-0.451794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-r2-P1-cre-3_at</th>\n",
       "      <td>0.140528</td>\n",
       "      <td>-0.502214</td>\n",
       "      <td>-0.163597</td>\n",
       "      <td>-0.283587</td>\n",
       "      <td>-0.389978</td>\n",
       "      <td>-0.262267</td>\n",
       "      <td>-0.431953</td>\n",
       "      <td>0.192936</td>\n",
       "      <td>0.024424</td>\n",
       "      <td>-0.070731</td>\n",
       "      <td>...</td>\n",
       "      <td>0.043830</td>\n",
       "      <td>0.053573</td>\n",
       "      <td>0.079236</td>\n",
       "      <td>0.160550</td>\n",
       "      <td>0.194999</td>\n",
       "      <td>0.588087</td>\n",
       "      <td>0.531057</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.757402</td>\n",
       "      <td>-0.308466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-r2-P1-cre-5_at</th>\n",
       "      <td>0.148749</td>\n",
       "      <td>-0.525167</td>\n",
       "      <td>-0.168589</td>\n",
       "      <td>-0.551300</td>\n",
       "      <td>-0.528519</td>\n",
       "      <td>-0.424885</td>\n",
       "      <td>-0.565857</td>\n",
       "      <td>0.172512</td>\n",
       "      <td>-0.074769</td>\n",
       "      <td>-0.159806</td>\n",
       "      <td>...</td>\n",
       "      <td>0.187002</td>\n",
       "      <td>0.162793</td>\n",
       "      <td>0.219069</td>\n",
       "      <td>0.326735</td>\n",
       "      <td>0.355099</td>\n",
       "      <td>0.728331</td>\n",
       "      <td>0.690192</td>\n",
       "      <td>0.757402</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.507809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "      <td>0.037070</td>\n",
       "      <td>0.257226</td>\n",
       "      <td>0.065501</td>\n",
       "      <td>0.612591</td>\n",
       "      <td>0.256257</td>\n",
       "      <td>0.385145</td>\n",
       "      <td>0.343723</td>\n",
       "      <td>-0.143690</td>\n",
       "      <td>0.047398</td>\n",
       "      <td>-0.105111</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.188470</td>\n",
       "      <td>-0.149833</td>\n",
       "      <td>-0.201614</td>\n",
       "      <td>-0.280555</td>\n",
       "      <td>-0.303186</td>\n",
       "      <td>-0.456953</td>\n",
       "      <td>-0.451794</td>\n",
       "      <td>-0.308466</td>\n",
       "      <td>-0.507809</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19093 rows Ã— 19094 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       1294_at   1316_at   1431_at   1438_at   1487_at  \\\n",
       "1316_at              -0.196297  1.000000 -0.107147  0.297960  0.401575   \n",
       "1431_at               0.265514 -0.107147  1.000000 -0.026961 -0.093128   \n",
       "1438_at              -0.158850  0.297960 -0.026961  1.000000  0.444146   \n",
       "1487_at              -0.390742  0.401575 -0.093128  0.444146  1.000000   \n",
       "1552256_a_at         -0.322735  0.437606 -0.239341  0.296720  0.395359   \n",
       "...                        ...       ...       ...       ...       ...   \n",
       "AFFX-r2-Ec-bioD-3_at  0.506664 -0.342041  0.143749 -0.477992 -0.568344   \n",
       "AFFX-r2-Ec-bioD-5_at  0.513898 -0.267291  0.138784 -0.463958 -0.477054   \n",
       "AFFX-r2-P1-cre-3_at   0.140528 -0.502214 -0.163597 -0.283587 -0.389978   \n",
       "AFFX-r2-P1-cre-5_at   0.148749 -0.525167 -0.168589 -0.551300 -0.528519   \n",
       "AFFX-TrpnX-M_at       0.037070  0.257226  0.065501  0.612591  0.256257   \n",
       "\n",
       "                      1552256_a_at  1552257_a_at  1552258_at  1552263_at  \\\n",
       "1316_at                   0.437606      0.492194   -0.238358    0.287821   \n",
       "1431_at                  -0.239341     -0.246241    0.172745   -0.272039   \n",
       "1438_at                   0.296720      0.441981   -0.112081    0.111677   \n",
       "1487_at                   0.395359      0.616744   -0.144776    0.321502   \n",
       "1552256_a_at              1.000000      0.578383   -0.170316    0.220145   \n",
       "...                            ...           ...         ...         ...   \n",
       "AFFX-r2-Ec-bioD-3_at     -0.434607     -0.658133    0.429023   -0.184792   \n",
       "AFFX-r2-Ec-bioD-5_at     -0.359832     -0.594893    0.442233   -0.120095   \n",
       "AFFX-r2-P1-cre-3_at      -0.262267     -0.431953    0.192936    0.024424   \n",
       "AFFX-r2-P1-cre-5_at      -0.424885     -0.565857    0.172512   -0.074769   \n",
       "AFFX-TrpnX-M_at           0.385145      0.343723   -0.143690    0.047398   \n",
       "\n",
       "                      1552264_a_at  ...  AFFX-r2-Ec-bioB-3_at  \\\n",
       "1316_at                   0.222830  ...             -0.044619   \n",
       "1431_at                  -0.196795  ...              0.344679   \n",
       "1438_at                  -0.172533  ...             -0.219205   \n",
       "1487_at                   0.183275  ...             -0.335935   \n",
       "1552256_a_at              0.501943  ...             -0.275150   \n",
       "...                            ...  ...                   ...   \n",
       "AFFX-r2-Ec-bioD-3_at     -0.123851  ...              0.656057   \n",
       "AFFX-r2-Ec-bioD-5_at     -0.115855  ...              0.669071   \n",
       "AFFX-r2-P1-cre-3_at      -0.070731  ...              0.043830   \n",
       "AFFX-r2-P1-cre-5_at      -0.159806  ...              0.187002   \n",
       "AFFX-TrpnX-M_at          -0.105111  ...             -0.188470   \n",
       "\n",
       "                      AFFX-r2-Ec-bioB-5_at  AFFX-r2-Ec-bioB-M_at  \\\n",
       "1316_at                          -0.012882             -0.056788   \n",
       "1431_at                           0.318784              0.317479   \n",
       "1438_at                          -0.196298             -0.257506   \n",
       "1487_at                          -0.332841             -0.372503   \n",
       "1552256_a_at                     -0.208974             -0.255458   \n",
       "...                                    ...                   ...   \n",
       "AFFX-r2-Ec-bioD-3_at              0.632837              0.678121   \n",
       "AFFX-r2-Ec-bioD-5_at              0.642518              0.684320   \n",
       "AFFX-r2-P1-cre-3_at               0.053573              0.079236   \n",
       "AFFX-r2-P1-cre-5_at               0.162793              0.219069   \n",
       "AFFX-TrpnX-M_at                  -0.149833             -0.201614   \n",
       "\n",
       "                      AFFX-r2-Ec-bioC-3_at  AFFX-r2-Ec-bioC-5_at  \\\n",
       "1316_at                          -0.128206             -0.148492   \n",
       "1431_at                           0.332442              0.277273   \n",
       "1438_at                          -0.360003             -0.399195   \n",
       "1487_at                          -0.476981             -0.495966   \n",
       "1552256_a_at                     -0.359736             -0.365745   \n",
       "...                                    ...                   ...   \n",
       "AFFX-r2-Ec-bioD-3_at              0.757525              0.765996   \n",
       "AFFX-r2-Ec-bioD-5_at              0.754380              0.762137   \n",
       "AFFX-r2-P1-cre-3_at               0.160550              0.194999   \n",
       "AFFX-r2-P1-cre-5_at               0.326735              0.355099   \n",
       "AFFX-TrpnX-M_at                  -0.280555             -0.303186   \n",
       "\n",
       "                      AFFX-r2-Ec-bioD-3_at  AFFX-r2-Ec-bioD-5_at  \\\n",
       "1316_at                          -0.342041             -0.267291   \n",
       "1431_at                           0.143749              0.138784   \n",
       "1438_at                          -0.477992             -0.463958   \n",
       "1487_at                          -0.568344             -0.477054   \n",
       "1552256_a_at                     -0.434607             -0.359832   \n",
       "...                                    ...                   ...   \n",
       "AFFX-r2-Ec-bioD-3_at              1.000000              0.973461   \n",
       "AFFX-r2-Ec-bioD-5_at              0.973461              1.000000   \n",
       "AFFX-r2-P1-cre-3_at               0.588087              0.531057   \n",
       "AFFX-r2-P1-cre-5_at               0.728331              0.690192   \n",
       "AFFX-TrpnX-M_at                  -0.456953             -0.451794   \n",
       "\n",
       "                      AFFX-r2-P1-cre-3_at  AFFX-r2-P1-cre-5_at  \\\n",
       "1316_at                         -0.502214            -0.525167   \n",
       "1431_at                         -0.163597            -0.168589   \n",
       "1438_at                         -0.283587            -0.551300   \n",
       "1487_at                         -0.389978            -0.528519   \n",
       "1552256_a_at                    -0.262267            -0.424885   \n",
       "...                                   ...                  ...   \n",
       "AFFX-r2-Ec-bioD-3_at             0.588087             0.728331   \n",
       "AFFX-r2-Ec-bioD-5_at             0.531057             0.690192   \n",
       "AFFX-r2-P1-cre-3_at              1.000000             0.757402   \n",
       "AFFX-r2-P1-cre-5_at              0.757402             1.000000   \n",
       "AFFX-TrpnX-M_at                 -0.308466            -0.507809   \n",
       "\n",
       "                      AFFX-TrpnX-M_at  \n",
       "1316_at                      0.257226  \n",
       "1431_at                      0.065501  \n",
       "1438_at                      0.612591  \n",
       "1487_at                      0.256257  \n",
       "1552256_a_at                 0.385145  \n",
       "...                               ...  \n",
       "AFFX-r2-Ec-bioD-3_at        -0.456953  \n",
       "AFFX-r2-Ec-bioD-5_at        -0.451794  \n",
       "AFFX-r2-P1-cre-3_at         -0.308466  \n",
       "AFFX-r2-P1-cre-5_at         -0.507809  \n",
       "AFFX-TrpnX-M_at              1.000000  \n",
       "\n",
       "[19093 rows x 19094 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.drop('1294_at')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
