{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samples</th>\n",
       "      <th>type</th>\n",
       "      <th>1007_s_at</th>\n",
       "      <th>1053_at</th>\n",
       "      <th>117_at</th>\n",
       "      <th>121_at</th>\n",
       "      <th>1255_g_at</th>\n",
       "      <th>1294_at</th>\n",
       "      <th>1316_at</th>\n",
       "      <th>1320_at</th>\n",
       "      <th>...</th>\n",
       "      <th>AFFX-r2-Ec-bioD-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioD-5_at</th>\n",
       "      <th>AFFX-r2-P1-cre-3_at</th>\n",
       "      <th>AFFX-r2-P1-cre-5_at</th>\n",
       "      <th>AFFX-ThrX-3_at</th>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>136</td>\n",
       "      <td>tumoral</td>\n",
       "      <td>8.767514</td>\n",
       "      <td>4.092134</td>\n",
       "      <td>6.006339</td>\n",
       "      <td>6.091951</td>\n",
       "      <td>2.738641</td>\n",
       "      <td>7.707960</td>\n",
       "      <td>6.702969</td>\n",
       "      <td>5.203349</td>\n",
       "      <td>...</td>\n",
       "      <td>13.852908</td>\n",
       "      <td>13.854510</td>\n",
       "      <td>14.953006</td>\n",
       "      <td>14.824910</td>\n",
       "      <td>3.710845</td>\n",
       "      <td>2.493435</td>\n",
       "      <td>2.913242</td>\n",
       "      <td>2.360155</td>\n",
       "      <td>2.758725</td>\n",
       "      <td>2.779035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>137</td>\n",
       "      <td>tumoral</td>\n",
       "      <td>9.177852</td>\n",
       "      <td>5.688045</td>\n",
       "      <td>5.599253</td>\n",
       "      <td>5.943833</td>\n",
       "      <td>2.667376</td>\n",
       "      <td>6.737484</td>\n",
       "      <td>7.336429</td>\n",
       "      <td>6.246692</td>\n",
       "      <td>...</td>\n",
       "      <td>13.784316</td>\n",
       "      <td>13.776588</td>\n",
       "      <td>14.944678</td>\n",
       "      <td>14.798245</td>\n",
       "      <td>3.840439</td>\n",
       "      <td>2.452855</td>\n",
       "      <td>2.657462</td>\n",
       "      <td>2.480106</td>\n",
       "      <td>2.811486</td>\n",
       "      <td>3.098742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>138</td>\n",
       "      <td>tumoral</td>\n",
       "      <td>8.415823</td>\n",
       "      <td>4.579540</td>\n",
       "      <td>5.562391</td>\n",
       "      <td>6.185697</td>\n",
       "      <td>2.747484</td>\n",
       "      <td>6.961582</td>\n",
       "      <td>6.809795</td>\n",
       "      <td>7.158062</td>\n",
       "      <td>...</td>\n",
       "      <td>13.833406</td>\n",
       "      <td>13.853554</td>\n",
       "      <td>14.929317</td>\n",
       "      <td>14.749283</td>\n",
       "      <td>3.960969</td>\n",
       "      <td>2.593110</td>\n",
       "      <td>2.839218</td>\n",
       "      <td>2.483863</td>\n",
       "      <td>2.939876</td>\n",
       "      <td>2.761388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>139</td>\n",
       "      <td>tumoral</td>\n",
       "      <td>8.943121</td>\n",
       "      <td>6.577537</td>\n",
       "      <td>5.850425</td>\n",
       "      <td>5.846352</td>\n",
       "      <td>2.429107</td>\n",
       "      <td>6.492797</td>\n",
       "      <td>7.594337</td>\n",
       "      <td>6.060994</td>\n",
       "      <td>...</td>\n",
       "      <td>13.601130</td>\n",
       "      <td>13.604113</td>\n",
       "      <td>14.927910</td>\n",
       "      <td>14.753407</td>\n",
       "      <td>3.821295</td>\n",
       "      <td>2.629036</td>\n",
       "      <td>2.798198</td>\n",
       "      <td>2.487149</td>\n",
       "      <td>2.652175</td>\n",
       "      <td>2.600322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>141</td>\n",
       "      <td>tumoral</td>\n",
       "      <td>8.658734</td>\n",
       "      <td>4.624722</td>\n",
       "      <td>6.882949</td>\n",
       "      <td>5.880421</td>\n",
       "      <td>2.517857</td>\n",
       "      <td>7.519646</td>\n",
       "      <td>6.840436</td>\n",
       "      <td>6.290670</td>\n",
       "      <td>...</td>\n",
       "      <td>13.961861</td>\n",
       "      <td>13.933318</td>\n",
       "      <td>14.994563</td>\n",
       "      <td>14.803339</td>\n",
       "      <td>3.990229</td>\n",
       "      <td>2.862215</td>\n",
       "      <td>2.766505</td>\n",
       "      <td>2.229823</td>\n",
       "      <td>2.630706</td>\n",
       "      <td>2.588111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 54677 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   samples     type  1007_s_at   1053_at    117_at    121_at  1255_g_at  \\\n",
       "0      136  tumoral   8.767514  4.092134  6.006339  6.091951   2.738641   \n",
       "1      137  tumoral   9.177852  5.688045  5.599253  5.943833   2.667376   \n",
       "2      138  tumoral   8.415823  4.579540  5.562391  6.185697   2.747484   \n",
       "3      139  tumoral   8.943121  6.577537  5.850425  5.846352   2.429107   \n",
       "4      141  tumoral   8.658734  4.624722  6.882949  5.880421   2.517857   \n",
       "\n",
       "    1294_at   1316_at   1320_at  ...  AFFX-r2-Ec-bioD-3_at  \\\n",
       "0  7.707960  6.702969  5.203349  ...             13.852908   \n",
       "1  6.737484  7.336429  6.246692  ...             13.784316   \n",
       "2  6.961582  6.809795  7.158062  ...             13.833406   \n",
       "3  6.492797  7.594337  6.060994  ...             13.601130   \n",
       "4  7.519646  6.840436  6.290670  ...             13.961861   \n",
       "\n",
       "   AFFX-r2-Ec-bioD-5_at  AFFX-r2-P1-cre-3_at  AFFX-r2-P1-cre-5_at  \\\n",
       "0             13.854510            14.953006            14.824910   \n",
       "1             13.776588            14.944678            14.798245   \n",
       "2             13.853554            14.929317            14.749283   \n",
       "3             13.604113            14.927910            14.753407   \n",
       "4             13.933318            14.994563            14.803339   \n",
       "\n",
       "   AFFX-ThrX-3_at  AFFX-ThrX-5_at  AFFX-ThrX-M_at  AFFX-TrpnX-3_at  \\\n",
       "0        3.710845        2.493435        2.913242         2.360155   \n",
       "1        3.840439        2.452855        2.657462         2.480106   \n",
       "2        3.960969        2.593110        2.839218         2.483863   \n",
       "3        3.821295        2.629036        2.798198         2.487149   \n",
       "4        3.990229        2.862215        2.766505         2.229823   \n",
       "\n",
       "   AFFX-TrpnX-5_at  AFFX-TrpnX-M_at  \n",
       "0         2.758725         2.779035  \n",
       "1         2.811486         3.098742  \n",
       "2         2.939876         2.761388  \n",
       "3         2.652175         2.600322  \n",
       "4         2.630706         2.588111  \n",
       "\n",
       "[5 rows x 54677 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "main = pd.read_csv('../Raw_Data/Prostate_GSE46602.csv')\n",
    "main.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting functions.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile functions.py\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scipy.stats as st\n",
    "from sklearn.model_selection import train_test_split\n",
    "from imblearn.under_sampling import RandomUnderSampler\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay\n",
    "from sklearn.preprocessing import QuantileTransformer\n",
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, cohen_kappa_score\n",
    "\n",
    "def DropSamples(df:pd.DataFrame):\n",
    "    if 'samples' in df.columns:\n",
    "        df = df.drop('samples', axis=1)\n",
    "    return df\n",
    "\n",
    "\n",
    "def calculate_difference(df:pd.DataFrame, tumor_column = 'type'):\n",
    "    \"\"\"\n",
    "    Calculate the mean difference between two groups in a DataFrame based on the specified 'tumor_column'.\n",
    "\n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): Input DataFrame containing data for comparison.\n",
    "    - tumor_column (str, optional): The column defining the groups. Default is 'type'.\n",
    "\n",
    "    Returns:\n",
    "    - pd.DataFrame: A new DataFrame containing the means of each column for each group, \n",
    "                    as well as the differences between the means of the two groups.\n",
    "    \"\"\"    \n",
    "    df2 = DropSamples(df).copy()\n",
    "\n",
    "    means_df = df2.groupby(tumor_column).agg('mean').reset_index()\n",
    "\n",
    "    # define Min_values Dict\n",
    "    min_values = {}\n",
    "    for col in means_df.columns[1:]:\n",
    "        min_values[col] = means_df[col][0] - means_df[col][1]\n",
    "\n",
    "    min_values_df = pd.DataFrame([min_values])\n",
    "    min_values_df[tumor_column] = 'difference'\n",
    "\n",
    "    # return new concatted dataframe\n",
    "    return pd.concat([means_df, min_values_df], axis=0)\n",
    "\n",
    "\n",
    "def plotSTDEV(df:pd.DataFrame, tumor_value = 'tumoral', tumor_column = 'type'):\n",
    "    \"\"\"\n",
    "    Plot the standard deviation (STDEV) distribution of a specified 'tumor_value' within a DataFrame,\n",
    "    grouped by the 'tumor_column'. The function generates a boxplot and a histogram for visualizing the distribution.\n",
    "\n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): Input DataFrame containing data for visualization.\n",
    "    - tumor_value (str, optional): The specific value within the 'tumor_column' for which the STDEV is plotted.\n",
    "                                  Default is 'tumoral'.\n",
    "    \"\"\"\n",
    "    df2 = DropSamples(df).copy()\n",
    "\n",
    "    sted_df = df2.groupby(tumor_column).std().reset_index()\n",
    "    sted_df_melt = sted_df[sted_df[tumor_column] == tumor_value].drop(tumor_column, axis=1).melt()\n",
    "\n",
    "    f, ax  = plt.subplots(2,1,figsize=(10,5), sharex=True)\n",
    "    ax[0] = sns.boxplot(sted_df_melt, ax=ax[0], orient='h')\n",
    "    ax[1] = sns.histplot(sted_df_melt, ax=ax[1])\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "    return sted_df_melt\n",
    "\n",
    "\n",
    "# make function to plot the thing that we talked about\n",
    "\n",
    "def getLowSTEDVList(df:pd.DataFrame, tumor_value = 'tumoral', tumor_column = 'type'):\n",
    "    \"\"\"\n",
    "    Get a list of variables with a standard deviation (STDEV) less than 1 for a specific tumor group in a DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): Input DataFrame containing data for analysis.\n",
    "    - tumor_value (str, optional): The specific tumor group to analyze. Default is 'tumoral'.\n",
    "    - tumor_column (str, optional): The column defining the tumor groups. Default is 'type'.\n",
    "\n",
    "    Returns:\n",
    "    - list: A list of variables with STDEV values less than 1 for the specified tumor group.\n",
    "    \"\"\"\n",
    "    df2 = DropSamples(df).copy()\n",
    "\n",
    "    sted_df = df2.groupby(tumor_column).std().reset_index()\n",
    "    sted_df_melt = sted_df[sted_df[tumor_column] == tumor_value].drop(tumor_column, axis=1).melt()\n",
    "\n",
    "    return sted_df_melt[sted_df_melt['value']<1]['variable'].to_list()\n",
    "\n",
    "\n",
    "def getListDifferentialGenes(df:pd.DataFrame, psig=0.05, tumor_value = 'tumoral', tumor_column = 'type'):\n",
    "    \"\"\"\n",
    "    Identify differentially expressed genes between a specified tumor group and the rest of the samples in a DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): Input DataFrame containing gene expression data.\n",
    "    - psig (float, optional): Significance threshold for identifying differentially expressed genes. Default is 0.05.\n",
    "    - tumor_value (str, optional): The specific tumor group for which differential expression is assessed. Default is 'tumoral'.\n",
    "    - tumor_column (str, optional): The column defining the tumor groups. Default is 'type'.\n",
    "\n",
    "    Returns:\n",
    "    - list: A list of gene names that are differentially expressed in the specified tumor group based on a t-test.\n",
    "    \"\"\"    \n",
    "    df2 = df.copy()\n",
    "\n",
    "    df_cancer = df2[df2[tumor_column]==tumor_value]\n",
    "    df_normal = df2[df2[tumor_column]!=tumor_value]\n",
    "\n",
    "    differential_expressed_genes = []\n",
    "    for col in df_cancer.columns[2:]:\n",
    "        t, pvalue= st.ttest_ind(df_cancer[col],df_normal[col], equal_var = False, alternative = 'two-sided')\n",
    "        if pvalue < psig:\n",
    "            differential_expressed_genes.append(str(col))\n",
    "\n",
    "    return differential_expressed_genes\n",
    "\n",
    "\n",
    "def find_correlated_genes(df:pd.DataFrame, threshold=0.95):\n",
    "    \"\"\"\n",
    "    Identify and analyze genes with high correlation coefficients in a gene expression DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): Input DataFrame containing gene expression data.\n",
    "    - threshold (float, optional): The correlation coefficient threshold for identifying highly correlated genes. Default is 0.95.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple: A tuple containing three elements:\n",
    "        1. List: Information about correlated genes, including column names, correlation values, and index names.\n",
    "        2. List: Names of genes to be excluded to avoid redundancy.\n",
    "        3. pd.DataFrame: The correlation matrix of the input DataFrame.\n",
    "    \"\"\"\n",
    "    df2 = DropSamples(df).copy()\n",
    "    \n",
    "    r_values = df2.corr()\n",
    "    \n",
    "    correlated_genes_list = []\n",
    "    # Create a mask for values above the threshold\n",
    "    mask = (r_values.to_numpy() > threshold) & (r_values.index.to_numpy() != r_values.columns.to_numpy()[:, None])\n",
    "\n",
    "    # Extract the column and index names where the mask is True\n",
    "    correlated_columns, correlated_rows = np.where(mask)\n",
    "\n",
    "    for col, index in zip(r_values.columns[correlated_columns], r_values.index[correlated_rows]):\n",
    "        value = r_values.at[index, col]\n",
    "        correlated_genes_list.append([col, value, index])\n",
    "\n",
    "    exclude_list = []\n",
    "    included_columns = []\n",
    "\n",
    "    for col in correlated_genes_list:\n",
    "        if col[0] not in included_columns:\n",
    "            exclude_list.append(col[0])\n",
    "            included_columns.append(col[2])\n",
    "\n",
    "    return correlated_genes_list, exclude_list, r_values\n",
    "\n",
    "def trainLogisticModel(df: pd.DataFrame, ColumnToPredict='type', tumor_value = 'tumoral'):\n",
    "    \"\"\"\n",
    "    Train a logistic regression model to predict a specified column in a DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): Input DataFrame containing data for training and testing the model.\n",
    "    - ColumnToPredict (str, optional): The column to be predicted using logistic regression. Default is 'type'.\n",
    "    - tumor_value (str, optional): The specific tumor group to be treated as the positive label. Default is 'tumoral'.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple: A tuple containing three elements:\n",
    "        1. LogisticRegression: Trained logistic regression model.\n",
    "        2. pd.DataFrame: Confusion matrix for model evaluation.\n",
    "        3. dict: Dictionary containing various classification scores (e.g., accuracy, precision, recall).\n",
    "    \"\"\"\n",
    "    df2 = DropSamples(df).copy()\n",
    "\n",
    "    # Split into X/Y based on ColumnToPredict\n",
    "    y = df2[ColumnToPredict]\n",
    "    X = df2.drop([ColumnToPredict], axis=1)\n",
    "    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=1337)\n",
    "\n",
    "    # Peform underSampler if values are not equal\n",
    "    if df2[ColumnToPredict].value_counts()[0] != df2[ColumnToPredict].value_counts()[1]:\n",
    "        rus = RandomUnderSampler(random_state=69)\n",
    "        X_train, y_train = rus.fit_resample(X_train, y_train)\n",
    "\n",
    "    # Create LogisticRegression model\n",
    "    classification = LogisticRegression(random_state=0, solver='lbfgs')\n",
    "    classification.fit(X_train, y_train)\n",
    "\n",
    "    # Create Prediction values\n",
    "    y_pred = classification.predict(X=X_test)\n",
    "\n",
    "    # return y_pred, y_test\n",
    "    model, cm, scores = LogisticModelView(model=classification, y_pred=y_pred, y_test=y_test.values, pos_label=tumor_value)\n",
    "    return model, cm, scores\n",
    "\n",
    "def predict(model: LogisticRegression, X, tumor_value = 'tumoral'):\n",
    "    \"\"\"\n",
    "    Do not use, not ready yet. Maybe will never be ready tbh might not really need this one :)\n",
    "    \"\"\"\n",
    "    model, cm, scores = LogisticModelView(model=model, y_pred=model.predict(X=X), y_test=X, pos_label=tumor_value)\n",
    "    return model, cm, scores\n",
    "\n",
    "def LogisticModelView(model: LogisticRegression, y_pred, y_test, pos_label):\n",
    "    \"\"\"\n",
    "    Evaluate and visualize the performance of a logistic regression model using a confusion matrix.\n",
    "\n",
    "    Parameters:\n",
    "    - model (LogisticRegression): Trained logistic regression model.\n",
    "    - y_pred: Predicted values from the model.\n",
    "    - y_test: True labels for evaluation.\n",
    "    - pos_label (str): The positive label, typically the specific tumor group.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple: A tuple containing three elements:\n",
    "        1. LogisticRegression: Trained logistic regression model.\n",
    "        2. pd.DataFrame: Confusion matrix for model evaluation.\n",
    "        3. dict: Dictionary containing various classification scores (e.g., accuracy, precision, recall).\n",
    "    \"\"\"\n",
    "    # Create Confusion Matrix\n",
    "    cm = confusion_matrix(y_test, y_pred)\n",
    "\n",
    "    disp = ConfusionMatrixDisplay(cm, display_labels=model.classes_)\n",
    "    disp.plot()\n",
    "\n",
    "    # Print Precision of CM\n",
    "    scores = {\n",
    "        'accuracy': accuracy_score(y_test, y_pred),\n",
    "        'precision': precision_score(y_true=y_test, y_pred=y_pred, pos_label=pos_label),\n",
    "        'recall': recall_score(y_test, y_pred, pos_label=pos_label),\n",
    "        'f1': f1_score(y_test, y_pred, pos_label=pos_label),\n",
    "        'kappa': cohen_kappa_score(y_test, y_pred)\n",
    "    }\n",
    "\n",
    "    return model, cm, scores\n",
    "\n",
    "\n",
    "\n",
    "def quantileTransformer(df:pd.DataFrame, qt = None):\n",
    "    \"\"\"\n",
    "    Apply quantile transformation to a DataFrame, transforming numerical columns to follow a normal distribution.\n",
    "\n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): Input DataFrame containing numerical data for transformation.\n",
    "    - qt (QuantileTransformer or None, optional): An existing QuantileTransformer object to apply, or None to fit a new one. Default is None.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple: A tuple containing two elements:\n",
    "        1. pd.DataFrame: Transformed DataFrame with numerical columns following a normal distribution.\n",
    "        2. QuantileTransformer: The fitted or provided QuantileTransformer object.\n",
    "    \"\"\"\n",
    "    df_transformed = df.copy()\n",
    "\n",
    "    if qt is None:\n",
    "        # Fit new dataset, if none is given\n",
    "        qt = QuantileTransformer (output_distribution ='normal')\n",
    "        qt.fit(df_transformed[df_transformed.columns[2:]])\n",
    "    \n",
    "    # Apply Transformation to data set\n",
    "    df_transformed[df_transformed.columns[2:]] = qt.transform(df_transformed[df_transformed.columns[2:]])\n",
    "    return df_transformed, qt\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
